In [1]:
%reload_ext autoreload
%autoreload 2
%matplotlib inline

# Задание №4 - Минимизация квадратичной функции.

## Начальная инициализация.

In [2]:
from sympy import *
import numpy as np
from scipy.spatial import distance

Объявляем функцию:  
 
$$\ f(x) = 2x^2 + 3.7y^2 + 4.7z^2 + xy - yz + xz + x - 2y + 3z + 7 $$

Следуя теории, наша квадратичная форма имеет вид:  
 
$$\ f(x) = \frac{1}{2}x^TAx+x^Tb $$  
 
Отсюда, наша матрица А и вектор b имеют вид: 
 
$ A=\begin{pmatrix}4& 1& 1\\1& 7.4& -1\\1& -1& 9.4\end{pmatrix}$  
 

$  b = \begin{pmatrix}1& -2& 3\end{pmatrix}$  
 
Инициализируем их:

In [3]:
x, y, z = symbols('x, y, z')
f = 2*x*x + 3.7*y*y + 4.7*z*z + x*y - y*z + x*z + x - 2*y + 3*z + 7
A = np.array([
    [4, 1, 1],
    [1, 7.4, -1 ],
    [1, -1, 9.4]], dtype=np.float64 )
b = np.array([1, -2, 3], dtype=np.float64)

### Метод наискорейшего градиентного спуска.

Сделаем начальную инициализацию параметров метода наискорейшего градиентного спуска.

In [15]:
x_k = np.array([-0.2, 0.2, -0.2], dtype=np.float64)
f_n = []  # list of function's values.
f_n.append(f.subs({x:x_k[0], y:x_k[1], z:x_k[2]}))
f_n

[6.25600000000000]

Как мы видим, значение функции в точке *`(-0.2, 0.2, -0.2)`* = **6.256**  
```
 
``` 
Здесь у нас переменные:
* x_k = рассматриваемая точка.
* f_n - запоминает значения $f(x_k)$ на каждой итерации.
* q_k - градиент.
* m_k - шаг метода.

### Метод наискорейшего градиентного спуска.

In [16]:
while True:
    q_k = np.dot(A, x_k) + b
    m_k = np.linalg.norm(q_k)**2/(np.dot(q_k, np.dot(A, q_k)))
    x_k -= m_k * q_k
    f_n.append(f.subs({x:x_k[0], y:x_k[1], z:x_k[2]}))
    if np.linalg.norm(np.dot(A, x_k) + b)/2 < 1e-6:
        break

In [17]:
f_n

[6.25600000000000,
 6.21231858695210,
 6.21031172454482,
 6.21010002469932,
 6.21007512805621,
 6.21007017301839,
 6.21006916426142,
 6.21006895531707,
 6.21006891203182,
 6.21006890306387,
 6.21006890120587,
 6.21006890082092,
 6.21006890074117,
 6.21006890072464,
 6.21006890072122,
 6.21006890072051]

In [18]:
len(f_n)

16

In [19]:
x_k

array([-0.25117417,  0.26855615, -0.26385857])

$\min\limits_{f(x)} = 6.21006890072051$ с точностью $\frac{1}{\delta} \begin{Vmatrix}Ax^k+b\end{Vmatrix}_2 < \frac{1}{10^6}$ был найден за **15** итераций. 

Точка $min = (-0.25117417,\space  0.26855615,\space -0.26385857)$.

### Метод наискорейшего покоординатного спуска.

Выполним начальную инициализацию:

In [20]:
x_k = np.array([-0.2, 0.2, -0.2], dtype=np.float64)
f_n = []
i = 1
f_n.append(f.subs({x:x_k[0], y:x_k[1], z:x_k[2]}))
f_n

[6.25600000000000]

Назначение переменных не менялось.

In [21]:
def find_ort(n):
    e = np.identity(3)  # Создаем единичную матрицу 3х3
    return e[n%3-1]

In [22]:
while True:
    e = find_ort(i)
    m_k = np.dot(e, (np.dot(A, x_k) + b))/np.dot(e, np.dot(A, e))
    x_k -= np.dot(m_k,e)
    f_n.append(f.subs({x:x_k[0], y:x_k[1], z:x_k[2]}))
    i+=1
    if np.linalg.norm(np.dot(A, x_k) + b)/2 < 1e-6:
        break

In [23]:
f_n

[6.25600000000000,
 6.25100000000000,
 6.22904729729730,
 6.21034426821877,
 6.21031996105232,
 6.21007998219730,
 6.21007887260145,
 6.21006975742631,
 6.21006929335499,
 6.21006896382617,
 6.21006891594015,
 6.21006890404829,
 6.21006890167054,
 6.21006890088677,
 6.21006890076594,
 6.21006890073132,
 6.21006890072244,
 6.21006890072087,
 6.21006890072046]

In [24]:
len(f_n)

19

In [25]:
x_k

array([-0.25117471,  0.2685563 , -0.2638584 ])

$\min\limits_{f(x)} = 6.21006890072046$ с точностью $\frac{1}{\delta} \begin{Vmatrix}Ax^k+b\end{Vmatrix}_2 < \frac{1}{10^6}$ был найден за **18** итерацию. 

Точка $min = (-0.25117471,\space  0.2685563,\space -0.2638584)$.